tests/core/editable/enterdiv.js

```diff
@@ -51,23 +86,47 @@ function testInsertHtml( htmlString ) {
 	};
 }
 
-function testGetData() {
-	return function( editor, bot ) {
-		bot.setData( '', function() {
-			var i = 0,
-				listener = editor.on( 'beforeGetData', function() {
-					++i;
-				} ),
-				spy = sinon.spy( CKEDITOR.editor.prototype, 'getData' ),
-				expectedGetDataCount = Number( editor.config.enterMode === CKEDITOR.ENTER_DIV );
+function generateEmptyParagraphsTests() {
+	var tags = [
+			'p',
+			'div',
+			'address',
+			'h1',
+			'h2',
+			'h3',
+			'h4',
+			'h5',
+			'h6',
+			'center',
+			'pre'
+		],
+		contents = [
+			'<br>',
+			'&nbsp;',
+			'\u00A0',
+			'&#160;'
+		];
 
-			editor.insertHtml( 'hublabubla' );
+	return CKEDITOR.tools.array.reduce( tags, function( tests, tag ) {
+		var tagTests = CKEDITOR.tools.array.reduce( contents, function( tests, content ) {
+			var test = {};
 
-			listener.removeListener();
-			spy.restore();
+			test[ 'empty paragraph test: ' + tag + ' with ' + CKEDITOR.tools.htmlEncode( content ) ] = function( editor ) {
+				var html = '<div>foo</div><div>bar</div>';
 
-			assert.areSame( expectedGetDataCount, spy.callCount, 'getData count' );
-			assert.areSame( 0, i, 'beforeGetData count' );
-		} );
-	};
+				editor.editable().setHtml( generateEmptyParagraph( tag, content ) );
+				editor.insertHtml( html );
+
+				assert.areSame( html, editor.getData() );
+			};
+
+			return CKEDITOR.tools.object.merge( tests, test );
+		}, {} );
+
+		return CKEDITOR.tools.object.merge( tests, tagTests );
+	}, {} );
```

Comment:
Maybe this test generation code would be a bit more readable if we would extract the cartesian product as a separate helper function? So we could just write `cartesianProduct( arr1, arr2 )` producing e.g. array-based tuple?

Reply:
After a second thought, it probably doesn't make much sense as the code is readable enough. Otherwise, we would have to avoid changing `tests` state directly which may complicate this test further.